dagshub/auth/tokens.py

```diff
@@ -0,0 +1,126 @@
+import datetime
+import logging
+import os
+import traceback
+from typing import Optional, Dict, List
+
+import yaml
+
+from dagshub.auth import oauth
+from dagshub.common import config
+
+logger = logging.getLogger(__name__)
+
+APP_TOKEN_TYPE = "app-token"
+
+
+class TokenStorage:
+    def __init__(self, cache_location: str = None, **kwargs):
+        if cache_location is None:
+            cache_location = config.cache_location
+        self.cache_location = cache_location
+        self.__token_cache: Optional[Dict[str, List[Dict]]] = None
+
+    @property
+    def _token_cache(self):
+        if self.__token_cache is None:
+            self.__token_cache = self._load_cache_file()
+        return self.__token_cache
+
+    def add_token(self, token: Dict, host: str = None):
+        if host is None:
+            host = config.host
+        if host not in self._token_cache:
+            self._token_cache[host] = []
+        self._token_cache[host].append(token)
+        self._store_cache_file()
+
+    def get_token(self, host: str = None):
+        if host is None:
+            host = config.host
+        tokens = self._token_cache.get(host, [])
+        app_tokens = list(filter(lambda t: t["token_type"] == APP_TOKEN_TYPE, tokens))
```

Comment:
I prefer list comprehension for a simple `==` comparison, but that's a matter of taste, you can leave it like that. 
Use `t.get("token_type")` though to prevent potential `KeyError` exceptions.